src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java

```diff
@@ -208,6 +215,40 @@ public boolean accept(Path path) {
     }
 
     /**
+     * Ensure reads in a pair fall in the same partition (input split), if the reads are queryname-sorted,
+     * so they are processed together. No shuffle is needed.
+     */
+    JavaRDD<GATKRead> putPairsInSamePartition(final SAMFileHeader header, final JavaRDD<GATKRead> reads) {
+        if (!header.getSortOrder().equals(SAMFileHeader.SortOrder.queryname)) {
+            return reads;
+        }
+        int numPartitions = reads.getNumPartitions();
+        // Find the first read in each partition
+        List<GATKRead> firstReadInEachPartition = reads
+                .mapPartitions((FlatMapFunction<Iterator<GATKRead>, GATKRead>) it -> Iterators.singletonIterator(it.next()))
+                .collect();
+        // Shift left, so that each partition will be joined with the first read from the _next_ partition
+        List<GATKRead> firstReadInNextPartition = new ArrayList<>(firstReadInEachPartition.subList(1, numPartitions));
+        firstReadInNextPartition.add(null); // the last partition does not have any reads to add to it
+
+        // Join the reads with the first read from the _next_ partition, then filter out the first and/or last read if not in a pair
+        return reads.zipPartitions(ctx.parallelize(firstReadInNextPartition, numPartitions),
+                (FlatMapFunction2<Iterator<GATKRead>, Iterator<GATKRead>, GATKRead>) (it1, it2) -> {
+            PeekingIterator<GATKRead> current = Iterators.peekingIterator(it1);
+            // skip the first read in the _current_ partition if it is the second in a pair since it will be handled in the previous partition
+            if (current.hasNext() && current.peek() != null && current.peek().isSecondOfPair()) {
+                current.next();
+            }
+            // append the first read in the _next_ partition to the _current_ partition if it is the second in a pair
+            PeekingIterator<GATKRead> next = Iterators.peekingIterator(it2);
+            if (next.hasNext() && next.peek() != null && next.peek().isSecondOfPair()) {
+                return Iterators.concat(current, next);
+            }
```

Comment:
It seems to me that we need to iterate until we find a read with a different name, since 1. the reads with the same name are not guaranteed to be sorted by "first of pair"/"second of pair" status, and 2. there can be supplementary alignments in addition to the primary alignments.